File: app/src/main/java/com/example/localai/data/network/Hex.java
package com.example.localai.data.network;

/** 十六进制编解码工具。 */
public final class Hex {

    private Hex() {
    }

    public static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(Character.forDigit((b >> 4) & 0xF, 16))
                    .append(Character.forDigit(b & 0xF, 16));
        }
        return sb.toString();
    }

    /** 严格解析：长度必须为偶数且全为十六进制字符，否则抛出 IllegalArgumentException。 */
    public static byte[] fromHex(String hex) {
        if (hex == null || (hex.length() & 1) != 0) {
            throw new IllegalArgumentException("bad hex length");
        }
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            int hi = digit(hex.charAt(2 * i));
            int lo = digit(hex.charAt(2 * i + 1));
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }

    private static int digit(char c) {
        if (c >= '0' && c <= '9') {
            return c - '0';
        }
        if (c >= 'a' && c <= 'f') {
            return c - 'a' + 10;
        }
        if (c >= 'A' && c <= 'F') {
            return c - 'A' + 10;
        }
        throw new IllegalArgumentException("bad hex char: " + c);
    }
}

File: app/src/main/kotlin/com/example/localai/data/network/Hex.kt
package com.example.localai.data.network

/** 十六进制编解码工具。 */
object Hex {

    @JvmStatic
    fun toHex(bytes: ByteArray): String {
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            sb.append(Character.forDigit((b.toInt() shr 4) and 0xF, 16))
                .append(Character.forDigit(b.toInt() and 0xF, 16))
        }
        return sb.toString()
    }

    /** 严格解析：长度必须为偶数且全为十六进制字符，否则抛出 IllegalArgumentException。 */
    @JvmStatic
    fun fromHex(hex: String): ByteArray {
        if (hex.length and 1 != 0) {
            throw IllegalArgumentException("bad hex length")
        }
        val out = ByteArray(hex.length / 2)
        for (i in out.indices) {
            val hi = digit(hex[2 * i])
            val lo = digit(hex[2 * i + 1])
            out[i] = ((hi shl 4) or lo).toByte()
        }
        return out
    }

    private fun digit(c: Char): Int {
        if (c in '0'..'9') {
            return c - '0'
        }
        if (c in 'a'..'f') {
            return c - 'a' + 10
        }
        if (c in 'A'..'F') {
            return c - 'A' + 10
        }
        throw IllegalArgumentException("bad hex char: $c")
    }
}
